changefeedccl: immediately stop sending webhook sink rows upon error

spiffyy99 opened the pull request:
Previously, the sink waited until flushing to acknowledge HTTP errors, leaving
any messages between the initial error and flush to potentially be out of
order. Now, errors are checked before each message is sent and the sink is
restarted if one is detected to maintain ordering.

Resolves #67772

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/67825)
<!-- Reviewable:end -->

miretskiy reviewed (changes requested):
> Reviewed 1 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67825)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @spiffyyeng)

___
*[pkg/ccl/changefeedccl/changefeed_test.go, line 4211 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67825#-Mf8QBQC8L3mVa30hFH5:-Mf8QBQC8L3mVa30hFH6:b-2k3jbf) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0820c7813350dedb026694a4ddb30e76a7d22a0/pkg/ccl/changefeedccl/changefeed_test.go#L4211)):*
> ```Go
> 		webhookFoo := foo.(*webhookFeed)
> 		// retry, then fail, then restart changefeed and successfully send messages
> 		webhookFoo.mockSink.SetStatusCodes(append(statusCodeRepeatedFromRetries(http.StatusInternalServerError),
> ```

can you break up long line?
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 255 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67825#-Mf8Qw3mAwbMwMHU1A4g:-Mf8Qw3mAwbMwMHU1A4h:bab7lhe) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0820c7813350dedb026694a4ddb30e76a7d22a0/pkg/ccl/changefeedccl/sink_webhook.go#L255)):*
> ```Go
> 			// shut down all other workers immediately if error encountered
> 			if err != nil {
> 				s.cancelFunc()
> ```

I think you need to add return statement here.

It's subtle, but basically, if you have multiple channels that are selectable (e.g. the Done() channel because you called cancelFunc(), and the caller that's trying to send a message), then which of the two cases will execute is undefined; thus it's possible that you will process an additional message, even after you "cancelled".
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 407 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67825#-Mf8RWSy3M_v-Qhz0yOY:-Mf8RWSz9QKLEgM-qU71:b3iu4l5) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0820c7813350dedb026694a4ddb30e76a7d22a0/pkg/ccl/changefeedccl/sink_webhook.go#L407)):*
<details><summary><i alt="summary">Quoted 6 lines of code…</i></summary>

> ```
> 	select {
> 	// non-blocking check for error, restart changefeed if encountered
> 	case <-s.inflight.errChan:
> 		return err
> 	default:
> 	}
> ```
</details>


This code is repeated many times now; we should move it to a helper... maybe
`s.inflight.hasError()` or some such.
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 416 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67825#-Mf8Sg2cBsGRn4ulXQRB:-Mf8Sg2cBsGRn4ulXQRC:b-fy2y7t) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0820c7813350dedb026694a4ddb30e76a7d22a0/pkg/ccl/changefeedccl/sink_webhook.go#L416)):*
> ```Go
> 	// check the webhook sink context in case workers have been terminated
> 	case <-s.ctx.Done():
> 		return s.ctx.Err()
> ```

I would find it more natural if we checked for errCh here.

nit: (not part of this PR per say, but I would appreciate if you could do this): I had to read this code few times, primarily because I was confused between ctx.Done and s.ctx.Done.
Could you please rename s.ctx and s.cancelFunc to `workerCtx` and `exitWorkers` respectively.
___
*[pkg/ccl/changefeedccl/sink_webhook_test.go, line 42 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67825#-Mf8Uz-m2-ELfgqBphGS:-Mf8Uz-nEJs6aQccerEf:b-blxdkl) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0820c7813350dedb026694a4ddb30e76a7d22a0/pkg/ccl/changefeedccl/sink_webhook_test.go#L42)):*
> ```Go
> // webhook sink should return for subsequent requests to ensure that an error
> // is reached even with the default retry behavior.
> func statusCodeRepeatedFromRetries(code int) []int {
> ```

nit: I would change this function to be
`func repeatStatusCode(code int, num int) []int`.

The reason is that I think it would be clearer from the call sites in the test if you had
`repeatStatusCode(StatusNotFound, defaultRetryConfig().MaxRetries + 1)`


<!-- Sent from Reviewable.io -->

spiffyy99 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67825)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy)

___
*[pkg/ccl/changefeedccl/changefeed_test.go, line 4211 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67825#-Mf8QBQC8L3mVa30hFH5:-Mf8qQXE72Gjbx_rUvbR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0820c7813350dedb026694a4ddb30e76a7d22a0/pkg/ccl/changefeedccl/changefeed_test.go#L4211)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

can you break up long line?
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 255 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67825#-Mf8Qw3mAwbMwMHU1A4g:-Mf8qRHO7ttJmJ13zFz4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0820c7813350dedb026694a4ddb30e76a7d22a0/pkg/ccl/changefeedccl/sink_webhook.go#L255)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I think you need to add return statement here.

It's subtle, but basically, if you have multiple channels that are selectable (e.g. the Done() channel because you called cancelFunc(), and the caller that's trying to send a message), then which of the two cases will execute is undefined; thus it's possible that you will process an additional message, even after you "cancelled".
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 407 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67825#-Mf8RWSy3M_v-Qhz0yOY:-Mf8qTXq0e5bQYU-LEzT:bvc7c91) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0820c7813350dedb026694a4ddb30e76a7d22a0/pkg/ccl/changefeedccl/sink_webhook.go#L407)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

> ```
> 	select {
> 	// non-blocking check for error, restart changefeed if encountered
> 	case <-s.inflight.errChan:
> 		return err
> 	default:
> 	}
> ```

This code is repeated many times now; we should move it to a helper... maybe
`s.inflight.hasError()` or some such.
</blockquote></details>

since we're moving it to the larger select statement i think it's fine to keep it as is (not sure how we'd insert a helper into the middle of the select)
___
*[pkg/ccl/changefeedccl/sink_webhook.go, line 416 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/67825#-Mf8Sg2cBsGRn4ulXQRB:-Mf8qT6U2uVucA7tFQ1d:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0820c7813350dedb026694a4ddb30e76a7d22a0/pkg/ccl/changefeedccl/sink_webhook.go#L416)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I would find it more natural if we checked for errCh here.

nit: (not part of this PR per say, but I would appreciate if you could do this): I had to read this code few times, primarily because I was confused between ctx.Done and s.ctx.Done.
Could you please rename s.ctx and s.cancelFunc to `workerCtx` and `exitWorkers` respectively.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

spiffyyeng commented:
side note: credit to @stevendanna for initially discovering this issue

spiffyy99 reviewed (commented):
> Dismissed @miretskiy from 4 discussions.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67825)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy)



<!-- Sent from Reviewable.io -->

spiffyy99 committed:
changefeedccl: immediately stop sending webhook sink rows upon error

Previously, the sink waited until flushing to acknowledge HTTP errors, leaving
any messages between the initial error and flush to potentially be out of
order. Now, errors are checked before each message is sent and the sink is
restarted if one is detected to maintain ordering.

Resolves #67772

Release note: None

miretskiy reviewed (commented):
> Reviewed 2 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67825)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy)



<!-- Sent from Reviewable.io -->

miretskiy reviewed (approved):
> Reviewed 1 of 3 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67825)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy)



<!-- Sent from Reviewable.io -->

spiffyyeng commented:
TFTR!

spiffyyeng commented:
bors r+
